packages/astro/src/core/render/README.md

```diff
@@ -0,0 +1,68 @@
+# `core/render`
+
+This directory contains most of Astro's high-level rendering APIs, including `renderPage`, `createRenderContext`, etc.
+
+- For rendering an Astro file, see [src/runtime/server/](../../runtime/server/).
+- For rendering an endpoint, see [src/core/endpoint/](../endpoint/).
+
+## Concepts
+
+The codebase has a few abstractions for rendering:
+
+### `RenderContext`
+
+Each render (page or endpoint) requires a `RenderContext` that contains information of:
+
+- The `Request` object
+- The matched `route`
+- The matched `pathname` (without the `base`)
+- The route `params` and `props`
+- Additional `styles`, `links`, `scripts`
+- And more!
+
+It is agnostic to what's being rendered and should only contain per-request information.
+
+### `Environment`
+
+Every app (dev and prod) has one `Environment`. It contains a subset of the Astro `settings`, `config`, and `routes` information needed for Astro's runtime to work.
+
+In dev, the `Environment` (aka `DevelopmentEnvironment`) is derived from the `settings`, `config`, and `routes` information directly as it's available then. In prod, it can only be derived from the `SSRManifest` as an intermediate layer so to keep the build output lean.
+
+The `Environment` should only contain the global state for all the requests.
```

Comment:
And then this can match the styling above:

```suggestion
**Permitted State:** The `Environment` should only contain the global state shared across all requests.
```